packages/wrangler/src/dev.ts

```diff
@@ -560,216 +700,154 @@ export async function startDev(args: StartDevOptions) {
 			args.config ||
 			(args.script && findWranglerConfig(path.dirname(args.script)));
 
-		// The ProxyWorker will have a stable host and port, so only listen for the first update
-		void devEnv.proxy.ready.promise.then(({ url }) => {
-			if (args.onReady) {
-				args.onReady(url.hostname, parseInt(url.port));
+		const authHook: AsyncHook<CfAccount, [Pick<Config, "account_id">]> = async (
+			config
+		) => {
+			const hotkeysDisplayed = !!unregisterHotKeys;
+			let accountId = args.accountId;
+			if (!accountId) {
+				unregisterHotKeys?.();
+				accountId = await requireAuth(config);
+				if (hotkeysDisplayed) {
+					assert(devEnv !== undefined);
+					unregisterHotKeys = registerDevHotKeys(
+						Array.isArray(devEnv) ? devEnv[0] : devEnv,
+						args
+					);
+				}
 			}
+			return {
+				accountId,
+				apiToken: requireApiToken(),
+			};
+		};
 
-			if (
-				(args.enableIpc || !args.onReady) &&
-				process.send &&
-				typeof vitest === "undefined"
-			) {
-				process.send(
-					JSON.stringify({
-						event: "DEV_SERVER_READY",
-						ip: url.hostname,
-						port: parseInt(url.port),
-					})
-				);
+		if (Array.isArray(configPath)) {
+			const runtime = new MultiworkerRuntimeController(configPath.length);
+
+			const primaryDevEnv = new DevEnv({ runtimes: [runtime] });
+
+			if (isInteractive() && args.showInteractiveDevSession !== false) {
+				unregisterHotKeys = registerDevHotKeys(primaryDevEnv, args);
 			}
-		});
 
-		if (!args.disableDevRegistry) {
-			teardownRegistryPromise = devRegistry((registry) =>
-				updateDevEnvRegistry(devEnv, registry)
+			// Set up the primary DevEnv (the one that the ProxyController will connect to)
```

Comment:
is there any reason we can't have a proxy controller + proxy worker per devEnv, and thus expose each worker over http?

Reply:
In theory, but I think it can be a follow up—this service binding approach will work for a lot of use cases. The ProxyWorker runs in it's own workerd instance, and it feels a bit weird to spin up loads of workerd instances for the proxies when the user workers all fit in one.

Reply:
ah I missed that they were in separate workerd instances. Makes sense to do in a followup, was just wondering :)